pkg/forwarder/failed_transaction_removal_policy.go

```diff
@@ -0,0 +1,138 @@
+}
+
+func newFailedTransactionRemovalPolicy(rootPath string, outdatedFileDayCount int) (*failedTransactionRemovalPolicy, error) {
+	if err := os.MkdirAll(rootPath, 0755); err != nil {
+		return nil, err
+	}
+
+	return &failedTransactionRemovalPolicy{
+		rootPath:           rootPath,
+		knownDomainFolders: make(map[string]struct{}),
+		outdatedFileTime:   time.Now().Add(time.Duration(-outdatedFileDayCount*24) * time.Hour),
+	}, nil
+}
+
+// RegisterDomain registers a domain name.
+func (p *failedTransactionRemovalPolicy) RegisterDomain(domainName string) (string, error) {
+	folder, err := p.getFolderPathForDomain(domainName)
+	if err != nil {
+		return "", err
+	}
+
+	p.knownDomainFolders[folder] = struct{}{}
+	return folder, nil
+}
+
+// RemoveOutdatedFiles removes the outdated files.
+// It removes files with extension retryTransactionsExtension:
+// - For domains not registered
+// - When a file is older than outDatedFileDayCount days
+func (p *failedTransactionRemovalPolicy) RemoveOutdatedFiles() ([]string, error) {
+	entries, err := ioutil.ReadDir(p.rootPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var filesRemoved []string
+	for _, domain := range entries {
+		if domain.Mode().IsDir() {
+			folderPath := path.Join(p.rootPath, domain.Name())
+			var files []string
+			if _, found := p.knownDomainFolders[folderPath]; found {
+				files, err = p.removeOutdatedFiles(folderPath)
+			} else {
+				files, err = p.removeUnknownDomain(folderPath)
+			}
+			if err != nil {
+				return nil, err
+			}
+			filesRemoved = append(filesRemoved, files...)
+		}
+	}
+	return filesRemoved, nil
+}
+
+func (p *failedTransactionRemovalPolicy) getFolderPathForDomain(domainName string) (string, error) {
+	// Use md5 for the folder name as the domainName is an url which can contain invalid charaters for a file path.
+	h := md5.New()
+	if _, err := io.WriteString(h, domainName); err != nil {
+		return "", err
+	}
+	folder := fmt.Sprintf("%x", h.Sum(nil))
+
+	return path.Join(p.rootPath, folder), nil
+}
+
+func (p *failedTransactionRemovalPolicy) removeUnknownDomain(folderPath string) ([]string, error) {
+	files, err := p.removeRetryFiles(folderPath, func(filename string) bool { return true })
+
+	// Try to remove the folder if it is empty
+	_ = os.Remove(folderPath)
+	return files, err
+}
+
+func (p *failedTransactionRemovalPolicy) removeOutdatedFiles(folderPath string) ([]string, error) {
+	return p.removeRetryFiles(folderPath, func(filename string) bool {
+		modTime, err := util.GetFileModTime(filename)
+		if err != nil {
+			return false
+		}
+		return modTime.Before(p.outdatedFileTime)
+	})
+}
+
+func (p *failedTransactionRemovalPolicy) removeRetryFiles(folderPath string, predicate func(string) bool) ([]string, error) {
+	files, err := p.getRetryFiles(folderPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var filesRemoved []string
+	for _, f := range files {
+		if predicate(f) {
+			if err = os.Remove(f); err != nil {
+				return nil, err
```

Comment:
Should we try to remove the rest of the file ?

Reply:
If the file cannot be removed then there is an high chance the other files cannot be removed. Do you think it worth handling this case?

Reply:
I **think** we should not block the removal logic for 1 file having the wrong permission for example. Doing as much as we can would be better. WDYT ?

Reply:
Code updated.